File: Robot Collisions/Robot Collisions.java
// Solution for Robot Collisions in JAVA

class Solution {
    public List<Integer> survivedRobotsHealths(int[] pos, int[] health, String dir) {
        int n = pos.length;
        Integer[] idx = new Integer[n];
        for (int i = 0; i < n; i++) idx[i] = i;

        Arrays.sort(idx, (a, b) -> pos[a] - pos[b]);

        Stack<Integer> st = new Stack<>();

        for (int i : idx) {
            if (dir.charAt(i) == 'R') st.push(i);
            else {
                while (!st.isEmpty() && health[i] > 0) {
                    int j = st.pop();
                    if (health[j] > health[i]) {
                        health[j]--;
                        health[i] = 0;
                        st.push(j);
                    } else if (health[j] < health[i]) {
                        health[i]--;
                        health[j] = 0;
                    } else {
                        health[i] = health[j] = 0;
                    }
                }
            }
        }

        List<Integer> res = new ArrayList<>();
        for (int h : health) if (h > 0) res.add(h);
        return res;
    }
}

File: Robot Collisions/Robot Collisions.py
# Solution for Robot Collisions in PY

class Solution:
    def survivedRobotsHealths(self, pos, health, dir):
        idx = sorted(range(len(pos)), key=lambda i: pos[i])
        stack = []

        for i in idx:
            if dir[i] == 'R':
                stack.append(i)
            else:
                while stack and health[i] > 0:
                    j = stack.pop()
                    if health[j] > health[i]:
                        health[j] -= 1
                        health[i] = 0
                        stack.append(j)
                    elif health[j] < health[i]:
                        health[i] -= 1
                        health[j] = 0
                    else:
                        health[i] = health[j] = 0

        return [h for h in health if h > 0]
